trunk/src/app/srs_app_conn.hpp

```diff
@@ -125,98 +126,68 @@ class SrsResourceManager : public ISrsCoroutineHandler, public ISrsResourceManag
-//
-// Usage for resource which manages itself in coroutine cycle, see SrsLazyGbSession:
-//      class Resource {
-//      private:
-//          SrsLazyObjectWrapper<Resource>* wrapper_;
-//      private:
-//          friend class SrsLazyObjectWrapper<Resource>;
-//          Resource(SrsLazyObjectWrapper<Resource>* wrapper) { wrapper_ = wrapper; }
-//      public:
-//          srs_error_t Resource::cycle() {
-//              srs_error_t err = do_cycle();
-//              _srs_gb_manager->remove(wrapper_);
-//              return err;
-//          }
-//      };
-//      SrsLazyObjectWrapper<Resource>* obj = new SrsLazyObjectWrapper<Resource>*();
-//      _srs_gb_manager->add(obj); // Add wrapper to resource manager.
-//      Start a coroutine to do obj->resource()->cycle().
+// This class implements the ISrsResource interface using a smart pointer, allowing the Manager to delete this
+// smart pointer resource, such as by implementing delayed release.
+// It embeds an SrsSharedPtr to provide the same interface, but it is not an inheritance relationship. Its usage
+// is identical to SrsSharedPtr, but they cannot replace each other. They are not related and cannot be converted
+// to one another.
+// Usage:
+//      SrsSharedResource<MyClass>* ptr = new SrsSharedResource<MyClass>(new MyClass());
+//      (*ptr)->do_something();
 //
-// Usage for resource managed by other object:
-//      class Resource {
-//      private:
-//          friend class SrsLazyObjectWrapper<Resource>;
-//          Resource(SrsLazyObjectWrapper<Resource>* /*wrapper*/) {
-//          }
-//      };
-//      class Manager {
-//      private:
-//          SrsLazyObjectWrapper<Resource>* wrapper_;
-//      public:
-//          Manager() { wrapper_ = new SrsLazyObjectWrapper<Resource>(); }
-//          ~Manager() { srs_freep(wrapper_); }
-//      };
-//      Manager* manager = new Manager();
-//      srs_freep(manager);
-//
-// Note that under-layer resource are destroyed by _srs_gc, which is literally equal to srs_freep. However, the root
-// wrapper might be managed by other resource manager, such as _srs_gb_manager for SrsLazyGbSession. Furthermore, other
-// copied out wrappers might be freed by srs_freep. All are ok, because all wrapper and resources are simply normal
-// object, so if you added to manager then you should use manager to remove it, and you can also directly delete it.
+//      ISrsResourceManager* manager = ...;
+//      manager->remove(ptr);
 template<typename T>
-class SrsLazyObjectWrapper : public ISrsResource
+class SrsSharedResource : virtual public ISrsResource
 {
 private:
-    T* resource_;
+    SrsSharedPtr<T> ptr_;
 public:
-    SrsLazyObjectWrapper() {
-        init(new T(this));
-    }
-    virtual ~SrsLazyObjectWrapper() {
-        resource_->gc_dispose();
-        if (resource_->gc_ref() == 0) {
-            _srs_gc->remove(resource_);
-        }
+    SrsSharedResource(T* ptr) : ptr_(ptr) {
```

Comment:
`SrsSharedResource(const SrsSharedPtr<T>& shared_ptr): ptr_(shared_ptr) {}`
I would like to add constructor with `SrsSharedPtr` also.

```
 SrsSharedPtr<T> shared_ptr = new SrsSharedPtr<MyClass>(new MyClass());
 SrsSharedResource<MyClass>* ptr = new SrsSharedResource<MyClass>(shared_ptr);
 (*ptr)->do_something();
```

Reply:
What is the use scenario?

Reply:
No such scenario yet, but missing this constructor make me felt unsafe in this case:

```
SrsSharedPtr<MyClass> shared_ptr(new MyClass());
SrsSharedResource<MyClass>* ptr = new SrsSharedResource<MyClass>(shared_ptr.get());
(*ptr)->do_something();
```

My top comment has errors, here is the fixed one.
```
 SrsSharedPtr<MyClass> shared_ptr(new MyClass());
 SrsSharedResource<MyClass>* ptr = new SrsSharedResource<MyClass>(shared_ptr);
 (*ptr)->do_something();
```

Reply:
Won't fix. Should never over design. Should never fix non-exists problem.